allennlp_rc/dataset_readers/transformer_squad.py

```diff
@@ -0,0 +1,325 @@
+            )
+            if any(i < 0 or i > len(tokenized_context_window) for i in window_token_answer_span):
+                # The answer is not contained in the window.
+                window_token_answer_span = None
+
+            if not self.skip_invalid_examples or window_token_answer_span is not None:
+                additional_metadata = {"id": qid}
+                instance = self.text_to_instance(
+                    question,
+                    tokenized_question,
+                    context,
+                    tokenized_context_window,
+                    answers,
+                    window_token_answer_span,
+                    additional_metadata,
+                )
+                yield instance
+
+            stride_start += space_for_context
+            if stride_start >= len(tokenized_context):
+                break
+            stride_start -= self.stride
+
+    @overrides
+    def text_to_instance(
+        self,  # type: ignore
+        question: str,
+        tokenized_question: List[Token],
+        context: str,
+        tokenized_context: List[Token],
+        answers: List[str],
+        token_answer_span: Optional[Tuple[int, int]],
+        additional_metadata: Dict[str, Any] = None,
+    ) -> Instance:
+        fields = {}
+
+        # make the question field
+        def cls_token():
+            return Token(
+                self._tokenizer.tokenizer.cls_token,
+                text_id=self._tokenizer.tokenizer.cls_token_id,
+                type_id=self.non_content_type_id,
+            )
+
+        def sep_token():
+            return Token(
+                self._tokenizer.tokenizer.sep_token,
+                text_id=self._tokenizer.tokenizer.sep_token_id,
+                type_id=self.non_content_type_id,
+            )
+
+        question_field = TextField(
+            (
+                [cls_token()]
+                + tokenized_question
+                + [sep_token(), sep_token()]
+                + tokenized_context
+                + [sep_token()]
+            ),
+            self._token_indexers,
+        )
+        fields["question"] = question_field
+        start_of_context = 1 + len(tokenized_question) + 2
+
+        # make the answer span
+        if token_answer_span is not None:
+            assert all(i >= 0 for i in token_answer_span)
+            assert token_answer_span[0] < token_answer_span[1]
+
+            fields["span_start"] = IndexField(
+                token_answer_span[0] + start_of_context, question_field
+            )
+            fields["span_end"] = IndexField(token_answer_span[1] + start_of_context, question_field)
+
+            assert all(
+                t.type_id == 0
+                for t in question_field.tokens[
+                    fields["span_start"].sequence_index : fields["span_end"].sequence_index
+                ]
+            )
+        else:
+            fields["span_start"] = IndexField(-1, question_field)
```

Comment:
Why not leave this empty?  `span_start` and `span_end` are optional in `forward`.

Reply:
What happens when we have a heterogeneous batch, where some instances have the span, and some do not?

My memory is fuzzy, but I think that's why I changed it to be like this.

Reply:
Ah, yeah, that would do it.  This wouldn't happen at training time, but it will at validation time.  Can you leave a comment saying why you're not just leaving the field empty?

Reply:
3a82f37f66c7d299a604c0f682768c046a30205f